contracts/src/v0.8/keystone/KeystoneFeedsConsumer.sol

```diff
@@ -2,27 +2,96 @@
+  event MessageReceived(bytes10 indexed workflowName, address indexed workflowOwner, uint256 nReports);
   event FeedReceived(bytes32 indexed feedId, uint256 price, uint64 timestamp);
 
-  constructor() {}
+  error UnauthorizedSender(address sender);
+  error UnauthorizedWorkflowOwner(address workflowOwner);
+  error UnauthorizedWorkflowName(bytes10 workflowName);
+
+  constructor() ConfirmedOwner(msg.sender) {}
 
   struct FeedReport {
-    bytes32 FeedID;
+    bytes32 FeedId;
     uint256 Price;
     uint64 Timestamp;
   }
 
-  function onReport(bytes32 workflowId, address workflowOwner, bytes calldata rawReport) external {
-    // TODO: validate sender and workflowOwner
+  mapping(bytes32 feedId => uint256 price) internal s_prices;
+  address[] internal allowedSenders;
+  address[] internal allowedWorkflowOwners;
+  bytes10[] internal allowedWorkflowNames;
+
+  function setConfig(address[] calldata _allowedSenders, address[] calldata _allowedWorkflowOwners, bytes10[] calldata _allowedWorkflowNames) external onlyOwner {
+    allowedSenders = _allowedSenders;
+    allowedWorkflowOwners = _allowedWorkflowOwners;
+    allowedWorkflowNames = _allowedWorkflowNames;
+  }
+
+  function onReport(bytes calldata metadata, bytes calldata rawReport) external {
+    bool allowed = false;
+    for (uint32 i = 0; i < allowedSenders.length; i++) {
+      if (msg.sender == allowedSenders[i]) {
+        allowed = true;
+        break;
+      }
+    }
+    if (!allowed) {
+      revert UnauthorizedSender(msg.sender);
+    }
+
+    (bytes10 workflowName, address workflowOwner) = _getInfo(metadata);
+    allowed = false;
+    for (uint32 i = 0; i < allowedWorkflowNames.length; i++) {
+      if (workflowName == allowedWorkflowNames[i]) {
+        allowed = true;
+        break;
+      }
+    }
+    if (!allowed) {
+      revert UnauthorizedWorkflowName(workflowName);
+    }
+
+    allowed = false;
+    for (uint32 i = 0; i < allowedWorkflowOwners.length; i++) {
+      if (workflowOwner == allowedWorkflowOwners[i]) {
+        allowed = true;
+        break;
+      }
+    }
+    if (!allowed) {
+      revert UnauthorizedWorkflowOwner(workflowOwner);
+    }
 
     FeedReport[] memory feeds = abi.decode(rawReport, (FeedReport[]));
     for (uint32 i = 0; i < feeds.length; i++) {
-      emit FeedReceived(feeds[i].FeedID, feeds[i].Price, feeds[i].Timestamp);
+      s_prices[feeds[i].FeedId] = feeds[i].Price;
+      emit FeedReceived(feeds[i].FeedId, feeds[i].Price, feeds[i].Timestamp);
+    }
+
+    emit MessageReceived(workflowName, workflowOwner, feeds.length);
+  }
+
+  function _getInfo(
+    bytes memory metadata
+  ) internal pure returns (bytes10 workflowName, address workflowOwner) {
+    // (first 32 bytes contain length of the byte array)
+    // workflow_cid             // offset 32, size 32
+    // workflow_name            // offset 64, size 10
+    // workflow_owner           // offset 74, size 20
+    // report_name              // offset 94, size  2
```

Comment:
This is 32 bytes total. You could return just the 32 bytes and do an equality comparison (even a hash is not needed).

Reply:
This might be tough when we allow multiple owners and/or names.